guide/src/free-threading.md

```diff
@@ -0,0 +1,165 @@
+Python interpreter runtime. See [PEP
+703](https://peps.python.org/pep-0703/#thread-states) for more background about
+how threads can be attached and detached from the interpreter runtime, in a
+manner analagous to releasing and acquiring the GIL in the GIL-enabled build.
+
+Calling into the CPython C API is only legal when an OS thread is explicitly
+attached to the interpreter runtime. In the GIL-enabled build, this happens when
+the GIL is acquired. In the free-threaded build there is no GIL, but the same C
+macros that release or acquire the GIL in the GIL-enabled build instead ask the
+interpreter to attach the thread to the Python runtime, and there can be many
+threads simultaneously attached.
+
+The main reason for attaching to the Python runtime is to interact with Python
+objects or call into the CPython C API. To interact with the Python runtime, the
+thread must register itself by attaching to the interpreter runtime. If you are
+not yet attached to the Python runtime, you can register the thread using the
+[`Python::with_gil`] function. Threads created via the Python `threading` module
+do not not need to do this, but all other OS threads that interact with the
+Python runtime must explicitly attach using `with_gil` and obtain a `'py`
+liftime.
+
+In the GIL-enabled build, PyO3 uses the `Python<'py>` type and the `'py` lifetime
+to signify that the global interpreter lock is held. In the freethreaded build,
+holding a `'py` lifetime means the thread is currently attached to the Python
+interpreter but other threads might be simultaneously interacting with the
+Python runtime.
+
+Since there is no GIL in the free-threaded build, releasing the GIL for
+long-running tasks is no longer necessary to ensure other threads run, but you
+should still detach from the interpreter runtime using [`Python::allow_threads`]
+when doing long-running tasks that do not require the CPython runtime. The
+garbage collector can only run if all threads are detached from the runtime (in
+a stop-the-world state), so detaching from the runtime allows freeing unused
+memory.
+
+## Runtime panics for multithreaded access of mutable `pyclass` instances
+
+If you wrote code that makes strong assumptions about the GIL protecting shared
+mutable state, it may not currently be straightforward to support free-threaded
+Python without the risk of runtime mutable borrow panics. PyO3 does not lock
+access to Python state, so if more than one thread tries to access a Python
+object that has already been mutably borrowed, only runtime checking enforces
+safety around mutably aliased Rust variables the Python interpreter can
+access. We believe that it would require adding an `unsafe impl` for `Send` or
+`Sync` to trigger this behavior in code using PyO3. Please report any issues
+related to runtime borrow checker errors on mutable pyclass implementations that
+do not make strong assumptions about the GIL.
```

Comment:
Sorry, why do we think that `unsafe impl` is required?

Reply:
I'm not actually sure, that came from @davidhewitt

Reply:
I think the point is that PyO3 uses `Send` and `Sync` bounds to rely on thread safety, but this is not really different from any other Rust code and also is relevant for everything, not just `#[pyclass]`.

Maybe we can replace this whole structure with something simpler like:

"`#[pyclass]` data is protected from concurrent access by a `RefCell`-like pattern of runtime borrow checking. Like a `RefCell`, PyO3 will raise exceptions (or in some cases panic) to enforce exclusive access for mutable borrows. With the GIL enabled it was unusual for these errors to be encountered in well structured code (using `allow_threads` or calling a `&self` `#[pymethod]` from a `&mut self` one could produce these errors)."

Reply:
Yes, that's much clearer. Perhaps we could even include an example of a Python snippet likely to fail? `[threading.Thread(target=some_obj.some_method_that_takes_self_mut).start() for _ in range(16)]` (but cleaned up)

Reply:
I added a code example.

One question: do the code examples in the docs get run as doctests? I modeled what I added based on another example in the guide that has some python code, but looking at the nox configuration I don't see anything that checks the code examples in the guide.

Reply:
That's a great question - you should add this new `.md` file to the list of `doctests!` in the bottom of `lib.rs` (around line 500).